File: 2501 Longest Square Streak in an Array/2501.py
class Solution:
    def longestSquareStreak(self, nums: List[int]) -> int:
        nums = sorted(set(nums))
        seen = [False] * 100001
        ans = -1

        for num in nums:
            seen[num] = True
        
        for num in nums:
            cnt = 1
            next_num = num * num

            while next_num <= nums[-1] and seen[next_num]:
                cnt += 1
                next_num *= next_num
            
            if cnt >= 2:
                ans = max(ans, cnt)    

        return ans

File: 2501 Longest Square Streak in an Array/2501.cpp
class Solution {
public:
    int longestSquareStreak(vector<int>& nums) {
        int ans = -1;
        sort(nums.begin(), nums.end());

        vector<bool> seen(100001, false);

        for (int num: nums){
            seen[num] = true;
        }

        int n = nums.size();

        for (int i = 0; i < n; i++){
            int cnt = 1;
            long long nextNum = (long long) nums[i] * nums[i];

            while (nextNum <= nums[n-1] && seen[nextNum]){
                cnt++;
                nextNum *= nextNum;
            }
            
            if (cnt >= 2){    
                ans = max(ans, cnt);
            }
        }

        return ans;
    }
};
